### PR TITLE
PLAT-396: return error when we try to call method on inactive object

### PR DESCRIPTION
**Pull request by ivanshibitov**
1. return error when we try to call method on inactive object
2. don't send states if object is Inactive 
3. add tests for that case

### CODE REVIEW

File: ledger-core/virtual/execute/execute.go [hunk 1]
```diff
@@ -227,13 +227,18 @@ func (s *SMExecute) stepWaitObjectReady(ctx smachine.ExecutionContext) smachine.
...
+	// }
+
 	if s.isConstructor {
 		switch objectState {
 		case object.Unknown:
 			panic(throw.Impossible())
 		case object.Inactive:
-			// attempt to create object that is deactivated :(
-			panic(throw.NotImplemented())
+			s.prepareExecutionError(throw.E("attempt to create object ( call constructor ) that is completely deactivated"))
```
Comment (bigbes) [inline ledger-core/virtual/execute/execute.go:240]: attempt to construct object that was deactivated

File: ledger-core/virtual/integration/deactivate_test.go [hunk 2]
```diff
@@ -139,3 +149,254 @@ func TestVirtual_DeactivateObject(t *testing.T) {
...
+
+					server.IncrementPulseAndWaitIdle(ctx)
+					Method_PrepareObject(ctx, server, payload.Inactive, object, prevPulse)
+
+					gotResult := make(chan struct{})
+
+					typedChecker := server.PublisherMock.SetTypedChecker(ctx, mc, server)
+					typedChecker.VCallResult.Set(func(res *payload.VCallResult) bool {
+
+						assert.Equal(t, res.Callee, object)
```
Comment (MariaVasilenko) [inline ledger-core/virtual/integration/deactivate_test.go:219]: please, fix Equal signature:
`func Equal(t TestingT, expected, actual interface{}, msgAndArgs ...interface{}) bool {`

File: ledger-core/virtual/integration/deactivate_test.go [hunk 3]
```diff
@@ -139,3 +149,254 @@ func TestVirtual_DeactivateObject(t *testing.T) {
...
+
+						return false // no resend msg
+					})
+
+					pl := payload.VCallRequest{
+						CallType:            callTypeTest.callType,
+						CallFlags:           payload.BuildCallFlags(contract.CallIntolerable, stateTest.objectState),
+						Caller:              server.GlobalCaller(),
+						Callee:              object,
+						CallSiteDeclaration: testwallet.GetClass(),
```
Comment (MariaVasilenko) [inline ledger-core/virtual/integration/deactivate_test.go:234]: use `class = gen.UniqueGlobalRef()` for CallSiteDeclaration field

File: ledger-core/virtual/integration/deactivate_test.go [hunk 4]
```diff
@@ -139,3 +149,254 @@ func TestVirtual_DeactivateObject(t *testing.T) {
...
+						CallSiteMethod:      "MyFavorMethod",
+						CallOutgoing:        reference.NewSelf(object.GetLocal()),
```
Comment (MariaVasilenko) [inline ledger-core/virtual/integration/deactivate_test.go:236]: use server.BuildRandomOutgoingWithPulse() for outgoing

File: ledger-core/virtual/execute/execute.go [hunk 5]
```diff
@@ -241,8 +241,8 @@ func (s *SMExecute) stepWaitObjectReady(ctx smachine.ExecutionContext) smachine.
 		case object.Unknown:
 			panic(throw.Impossible())
 		case object.Inactive:
-			// attempt to create object that is deactivated :(
-			panic(throw.Impossible())
+			s.prepareExecutionError(throw.E("attempt to construct object that was completely deactivated"))
```
Comment (aalekseev-ru) [inline ledger-core/virtual/execute/execute.go:244]: I am sure that it is impossible to call Constructor on Inactive object

File: ledger-core/virtual/execute/execute.go [hunk 6]
```diff
@@ -264,7 +264,8 @@ func (s *SMExecute) stepWaitObjectReady(ctx smachine.ExecutionContext) smachine.
 		}))
 		return ctx.Jump(s.stepSendCallResult)
 	case object.Inactive:
-		panic(throw.Impossible())
+		s.prepareExecutionError(throw.E("attempt to call method on object that is completely deactivated"))
```
Comment (aalekseev-ru) [inline ledger-core/virtual/execute/execute.go:267]: it's better for debugging to throw struct with object ref

File: ledger-core/virtual/integration/deactivate_test.go [hunk 7]
```diff
@@ -139,3 +149,254 @@ func TestVirtual_DeactivateObject(t *testing.T) {
...
+	server.IncrementPulseAndWaitIdle(ctx)
+	{
+		// Create object
+		Method_PrepareObject(ctx, server, payload.Ready, object, prevPulse)
+	}
+
+	isolation := contract.MethodIsolation{Interference: contract.CallTolerable, State: contract.CallDirty}
+	{
+		// execution mock for deactivation
+		descr := descriptor.NewObject(object, server.RandomLocalWithPulse(), server.RandomGlobalWithPulse(), makeRawWalletState(initialBalance), false)
```
Comment (MariaVasilenko) [inline ledger-core/virtual/integration/deactivate_test.go:287]: don't use wallet in integration tests

File: ledger-core/virtual/execute/execute.go (hunk 5 above)
Comment (ivanshibitov) [reply ledger-core/virtual/execute/execute.go:244]: leave as it was: panic(throw.Impossible())

File: ledger-core/virtual/integration/deactivate_test.go (hunk 7 above)
Comment (ivanshibitov) [reply ledger-core/virtual/integration/deactivate_test.go]: I understand, what you are talking about

File: ledger-core/virtual/integration/deactivate_test.go (hunk 4 above)
Comment (ivanshibitov) [reply ledger-core/virtual/integration/deactivate_test.go]: done

File: ledger-core/virtual/integration/deactivate_test.go (hunk 7 above)
Comment (MariaVasilenko) [reply ledger-core/virtual/integration/deactivate_test.go]: `makeRawWalletState(initialBalance)` -> `State: []byte("some memory")`

File: ledger-core/virtual/execute/execute.go [hunk 8]
```diff
@@ -241,7 +241,6 @@ func (s *SMExecute) stepWaitObjectReady(ctx smachine.ExecutionContext) smachine.
 		case object.Unknown:
 			panic(throw.Impossible())
 		case object.Inactive:
-			// attempt to create object that is deactivated :(
 			panic(throw.Impossible())
```
Comment (ruz) [inline ledger-core/virtual/execute/execute.go:244]: this is very possible situation

File: ledger-core/virtual/execute/execute.go [hunk 9]
```diff
@@ -264,7 +263,11 @@ func (s *SMExecute) stepWaitObjectReady(ctx smachine.ExecutionContext) smachine.
 		}))
 		return ctx.Jump(s.stepSendCallResult)
 	case object.Inactive:
-		panic(throw.Impossible())
+		s.prepareExecutionError(throw.E("attempt to call method on object that is completely deactivated", struct {
```
Comment (ruz) [inline ledger-core/virtual/execute/execute.go:266]: this is not correct place to check

File: ledger-core/virtual/execute/execute.go (hunk 8 above)
Comment (aalekseev-ru) [reply ledger-core/virtual/execute/execute.go:244]: example?

File: ledger-core/virtual/execute/execute.go (hunk 9 above)
Comment (ivanshibitov) [reply ledger-core/virtual/execute/execute.go:257]: It's kind of short path for checking. There is one more in 'stepStartRequestProcessing'

File: ledger-core/virtual/integration/deactivate_test.go [hunk 10]
```diff
@@ -388,7 +395,7 @@ func TestVirtual_CallMethod_On_DeactivatedDirtyState(t *testing.T) {
 					Arguments:           insolar.MustSerialize([]interface{}{}),
 				}
 				server.SendPayload(ctx, &pl)
-				commontestutils.WaitSignalsTimed(t, 10*time.Second, gotResult)
+				commontestutils.WaitSignalsTimed(t, 100*time.Second, gotResult)
```
Comment (aalekseev-ru) [inline ledger-core/virtual/integration/deactivate_test.go:398]: return 10 please

File: ledger-core/virtual/execute/execute.go (hunk 8 above)
Comment (ruz) [reply ledger-core/virtual/execute/execute.go:235]: duplicate request of constructor after deactivation in previous pulse. I wrote about that in slack.

File: ledger-core/virtual/execute/execute.go (hunk 9 above)
Comment (ruz) [reply ledger-core/virtual/execute/execute.go:257]: duplicate request of method after deactivation in previous pulse. I wrote about this in slack.

File: ledger-core/virtual/execute/execute.go [hunk 11]
```diff
@@ -541,21 +535,38 @@ func (s *SMExecute) getDescriptor(state *object.SharedState) descriptor.Object {
 func (s *SMExecute) stepStartRequestProcessing(ctx smachine.ExecutionContext) smachine.StateUpdate {
 	var (
 		objectDescriptor descriptor.Object
+		isDeactivated    bool
 	)
 	action := func(state *object.SharedState) {
+		objectDescriptor = s.getDescriptor(state)
+		if objectDescriptor != nil && objectDescriptor.Deactivated() {
+			if !state.KnownRequests.Delete(s.execution.Isolation.Interference, s.execution.Outgoing) {
```
Comment (ruz) [inline ledger-core/virtual/execute/execute.go:543]: this is not correct way. why do you Delete anything from the table?

Comment (ivanshibitov) [reply ledger-core/virtual/execute/execute.go:543]: Previously, we only added this reference, that is why I only remove this reference

Comment (ivanshibitov) [reply ledger-core/virtual/execute/execute.go:543]: What do you mean 'anything'?

Comment (ivanshibitov) [reply ledger-core/virtual/execute/execute.go]: I remove 'Delete'. And we store result of such execution

File: ledger-core/virtual/execute/execute.go (hunk 8 above)
Comment (ivanshibitov) [reply ledger-core/virtual/execute/execute.go]: move this check to stepStartRequestProcessing

File: ledger-core/virtual/execute/execute.go (hunk 9 above)
Comment (ivanshibitov) [reply ledger-core/virtual/execute/execute.go]: move check to stepStartRequestProcessing

**Pull request merged**